analytics_data_api/management/commands/generate_data.py

```diff
@@ -0,0 +1,339 @@
+# pylint: disable=line-too-long,invalid-name
+"""
+Abstract base management command meant to be called by other commands that generate fake data
+"""
+
+import abc
+import datetime
+import logging
+import math
+import random
+
+from django.conf import settings
+from django.core.management.base import BaseCommand
+from django.utils import timezone
+from tqdm import tqdm
+
+from analytics_data_api.constants import engagement_events
+from analytics_data_api.v0 import models
+from analyticsdataserver.clients import CourseBlocksApiClient
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+# http://stackoverflow.com/a/3590105
+def constrained_sum_sample_pos(num_values, total):
+    """Return a randomly chosen list of n positive integers summing to total.
+    Each such list is equally likely to occur."""
+
+    dividers = sorted(random.sample(range(1, total), num_values - 1))
+    return [a - b for a, b in zip(dividers + [total], [0] + dividers)]
+
+
+def get_count(start):
+    delta = 25 * random.gauss(0, 1)
+    return int(start + delta)
+
+
+class FakeDataCommand(BaseCommand, metaclass=abc.ABCMeta):
+
+    @abc.abstractmethod
+    def add_arguments(self, parser):
+        pass
+
+    def generate_daily_data(self, course_id, start_date, end_date, database,
+                            delete_data=True, add_birth_year=True, use_current_cumulative=False):
+        # Use the preset ratios below to generate data in the specified demographics
+
+        gender_ratios = {
+            'm': 0.6107,
+            'f': 0.3870,
+            'o': 0.23
+        }
+        education_level_ratios = {
+            'associates': 0.058,
+            'bachelors': 0.3355,
+            'primary': 0.0046,
+            'secondary': 0.2442,
+            'junior_secondary': 0.0286,
+            'masters': 0.2518,
+            'none': 0.0032,
+            'other': 0.0271,
+            'doctorate': 0.0470
+        }
+        country_ratios = {
+            'US': 0.33,
+            'GH': 0.12,
+            'IN': 0.10,
+            'CA': 0.14,
+            'CN': 0.22,
+            'DE': 0.08,
+            'UNKNOWN': 0.01
+        }
+
+        enrollment_mode_ratios = {
+            'audit': 0.15,
+            'credit': 0.15,
+            'honor': 0.25,
+            'professional': 0.10,
+            'verified': 0.35
+        }
```

Comment:
all of these ratios and the birth year ratios (which we could make all the time) could live outside the function to increase function readability